Refactor some fundamentals in ArgumentCodersCocoa.h

beidson opened the pull request:
#### 90ecb2cd9622645eadbe43708accc91290196bdf
<pre>
Refactor some fundamentals in ArgumentCodersCocoa.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=264385">https://bugs.webkit.org/show_bug.cgi?id=264385</a>
<a href="https://rdar.apple.com/118099847">rdar://118099847</a>

Reviewed by Alex Christensen.

- Make all encode() methods work in terms of a single master encode() method
- Get rid of many extra decode() methods
- Make all decode() methods work in terms of a single master decode() method
- Rename the final decode() method that takes &quot;allowed classes&quot; to &quot;decodeWithAllowedClasses&quot; for compile
  safety and clarity
- Move the main &quot;decodeWithAllowedClasses&quot; entry point to Decoder itself to avoid re-entrancy in some situations.
- Make the way that &quot;decodeWithAllowedClasses&quot; gets the default array-of-allowed-classes a template so
  our softlinked classes can work with the default parameter
- Update callers that used to explicitly pass an allowed class to rely on the default, when possible.

Why do all this?

It makes Obj-C decode use nicer to write and to look at.

But also upcoming work for NSArray and NSDictionary will completely upend how we encode/decode for our Obj-C types
and this patch is a standalone prerequisite to enabling that work.

* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::getClass):
(IPC::Decoder::decodeWithAllowedClasses):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::decode):
(IPC::ArgumentCoder&lt;SoftLinkedMember&gt;::decode):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::decodeWithAllowedClasses):
(IPC::ArgumentCoder&lt;RetainPtr&lt;T&gt;&gt;::decode):
(IPC::encode):
(IPC::decode): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeFontInternal):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::getClass&lt;PKPayment&gt;):
(IPC::getClass&lt;PKContact&gt;):
(IPC::getClass&lt;PKPaymentMerchantSession&gt;):
(IPC::getClass&lt;PKPaymentMethod&gt;):
(IPC::ArgumentCoder&lt;WebCore::Payment&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::PaymentContact&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::PaymentMerchantSession&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::PaymentMethod&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::PaymentSessionError&gt;::decode):
(IPC::getClass&lt;DDScannerResult&gt;):
(IPC::ArgumentCoder&lt;WebCore::DataDetectorElementInfo&gt;::decode):
(IPC::getClass&lt;AVOutputContext&gt;):
(IPC::ArgumentCoder&lt;WebCore::MediaPlaybackTargetContext&gt;::decodePlatformData):
(IPC::ArgumentCoder&lt;WebCore::TextRecognitionDataDetector&gt;::decodePlatformData):
(IPC::getClass&lt;VKCImageAnalysis&gt;):
(IPC::ArgumentCoder&lt;RetainPtr&lt;VKCImageAnalysis&gt;&gt;::decode):
* Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
(WebKit::ObjCObjectGraph::decode):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder&lt;WebCore::Credential&gt;::decodePlatformData):

Canonical link: <a href="https://commits.webkit.org/270461@main">https://commits.webkit.org/270461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d847459df1330503b36a21620ac336ebc40a886f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26831 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27650 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23414 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5934 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1586 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23561 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22023 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28232 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/25709 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/2750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29069 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23338 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26916 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2721 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/973 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4101 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6131 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3177 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3058 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
```diff
@@ -607,9 +627,14 @@
     encoder << info.elementBounds;
 }
 
+template<> Class getClass<DDScannerResult>()
```
It might be nice to put all these overloads together.

beidson replied:
I did it this way because each of them was in an #if ENABLE() block.

Will go lump all the PK ones together, though.

beidson committed:
Refactor some fundamentals in ArgumentCodersCocoa.h
https://bugs.webkit.org/show_bug.cgi?id=264385
rdar://118099847

Reviewed by Alex Christensen.

- Make all encode() methods work in terms of a single master encode() method
- Get rid of many extra decode() methods
- Make all decode() methods work in terms of a single master decode() method
- Rename the final decode() method that takes "allowed classes" to "decodeWithAllowedClasses" for compile
  safety and clarity
- Move the main "decodeWithAllowedClasses" entry point to Decoder itself to avoid re-entrancy in some situations.
- Make the way that "decodeWithAllowedClasses" gets the default array-of-allowed-classes a template so
  our softlinked classes can work with the default parameter
- Update callers that used to explicitly pass an allowed class to rely on the default, when possible.

Why do all this?

It makes Obj-C decode use nicer to write and to look at.

But also upcoming work for NSArray and NSDictionary will completely upend how we encode/decode for our Obj-C types
and this patch is a standalone prerequisite to enabling that work.

* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::getClass):
(IPC::Decoder::decodeWithAllowedClasses):
* Source/WebKit/Scripts/generate-serializers.py:
(decode_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::OtherClass>::decode):
(IPC::ArgumentCoder<SoftLinkedMember>::decode):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::decodeWithAllowedClasses):
(IPC::ArgumentCoder<RetainPtr<T>>::decode):
(IPC::encode):
(IPC::decode): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::decodeFontInternal):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::getClass<PKPayment>):
(IPC::getClass<PKContact>):
(IPC::getClass<PKPaymentMerchantSession>):
(IPC::getClass<PKPaymentMethod>):
(IPC::ArgumentCoder<WebCore::Payment>::decode):
(IPC::ArgumentCoder<WebCore::PaymentContact>::decode):
(IPC::ArgumentCoder<WebCore::PaymentMerchantSession>::decode):
(IPC::ArgumentCoder<WebCore::PaymentMethod>::decode):
(IPC::ArgumentCoder<WebCore::PaymentSessionError>::decode):
(IPC::getClass<DDScannerResult>):
(IPC::ArgumentCoder<WebCore::DataDetectorElementInfo>::decode):
(IPC::getClass<AVOutputContext>):
(IPC::ArgumentCoder<WebCore::MediaPlaybackTargetContext>::decodePlatformData):
(IPC::ArgumentCoder<WebCore::TextRecognitionDataDetector>::decodePlatformData):
(IPC::getClass<VKCImageAnalysis>):
(IPC::ArgumentCoder<RetainPtr<VKCImageAnalysis>>::decode):
* Source/WebKit/Shared/mac/ObjCObjectGraph.mm:
(WebKit::ObjCObjectGraph::decode):
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder<WebCore::Credential>::decodePlatformData):

Canonical link: https://commits.webkit.org/270461@main
